File: solutions/101. Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return root && isSymmetric_(root->left, root->right);
    }
​
private:
    /* This is a recursive approach. */
    bool isSymmetric_(TreeNode* left, TreeNode* right) {
        if (!isBothNullOrEqualVal_(left, right)) {
            return false;
        }
        if (isBothNull_(left, right)) {
            return true;
        }
        return isSymmetric_(left->left, right->right)
            && isSymmetric_(left->right, right->left);
    }
​
    bool isBothNullOrEqualVal_(TreeNode* n, TreeNode* m) {
        if (isBothNull_(n, m)) {
            return true;
        }
        if ((!n) || (!m)) {
            return false;
        }
        return n->val == m->val;
    }
​
    bool isBothNull_(TreeNode* n, TreeNode* m) {
        return (!n) && (!m);
    }
};
​

File: solutions/101. Symmetric Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Queue<TreeNode> queue = new LinkedList<>();
​
    /**
     * This is a iterative approach.
     * For example, a tree has the following structure:
     *       1     --- level 1 (won't be offered)
     *     2   3   --- level 2
     *   4  5 6  7 --- level 3
     * nodes will be offered in the order of 2 3 4 7 5 6,
     * they are checked in a symmetric breadth-first fashion.
     */
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return false;
        }
        queue.offer(root.left);
        queue.offer(root.right);
        while (!queue.isEmpty()) {
            final TreeNode n = queue.poll();
            final TreeNode m = queue.poll();
            if (isBothNull(n, m)) {
                continue;
            }
            if (isEqualVal(n, m)) {
                queue.offer(n.left);
                queue.offer(m.right);
​
                queue.offer(n.right);
                queue.offer(m.left);
            } else {
                return false;
            }
        }
        return true;
    }
​
    private boolean isBothNull(final TreeNode n, final TreeNode m) {
        return n == null && m == null;
    }
​
    private boolean isEqualVal(final TreeNode n, final TreeNode m) {
        if (isBothNull(n, m)) {  // for logic complete, though we've already checked above
            return true;
        }
        if (n == null || m == null) {
            return false;
        }
        return n.val == m.val;
    }
}
​
